### DISCUSSION TITLE
Order of ack for durable sessions

### DISCUSSION THREAD

**Discussion by jsobus-neurosys** [emqx/emqx#13683] [Q&A]
Hi guys,

We recently started thinking about using durable sessions for our events - we would like to use the feature because of possibility to retain messages on client during restart of a node.

We are having issues doing so, when we were not using a durable sessions we were making acks not in order - which allow us to use bigger throughput of messages. When we switched to durable sessions, without any change in our client code we started receiving errors like this on client side:

```
[0m 2024-08-22T14:40:58.572781+00:00 [warning] clientid: XxXxX_e99c576b8ef0, msg: out-of-order_commit, peername: 172.18.0.8:56966, username: XxXxX, expected: 3, seqno: 25, packet_id: 25, track: puback
[0m 2024-08-22T14:40:58.573395+00:00 [warning] clientid: XxXxX_e99c576b8ef0, msg: puback_packetId_not_found, peername: 172.18.0.8:56966, username: XxXxX, packetId: 25
```

Does this mean that using durable sessions we need to ack messages in order we received them? We have not found any information about it in the docs, so we are not sure how to proceed with our implementation.

**Comment by thalesmg** [comment]
Hi, perhaps you're observing those messages when clients reconnect and re-send `PUBLISH` packets with QoS > 0?

If so, they should re-send in the same order the orignal ones were sent.

From the [MQTT spec](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc384800478):

> When the Client re-sends any PUBLISH packets, it MUST re-send them in the order in which the original PUBLISH packets were sent (this applies to QoS 1 and QoS 2 messages) [MQTT-4.6.0-1]

**Reply by jsobus-neurosys** [reply]
@thalesmg thanks for a response!

Yes, we understand that this is not strictly as in MQTT spec but we found this discussion: https://github.com/emqx/emqx/discussions/4703 and like I said before we were working this way without durable sessions. As discussed it allows for great increase of throughput of messages, we are happy that EMQX allow this :)

Now when switched to durable sessions it looks it is strict on this MQTT spec and it work differently in this case? Or there is some other detail we are missing still allowing for concurrency of processing the messages?
